### PR TITLE
refResolver index.d.ts typescript typings

### PR DESCRIPTION
**Pull request by samzilverberg**
resolves https://github.com/fastify/fastify-swagger/issues/501

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by darkgl0w** [comment]
Can you add a unit test for this typescript definition please ?

**Comment by samzilverberg** [comment]
added test.

File: index.d.ts [hunk 1]
```diff
@@ -118,6 +118,10 @@ export interface FastifyDynamicSwaggerOptions extends FastifySwaggerOptions {
    * Overwrite the route schema
    */
   transform?: Function;
+
+  refResolver?: {
+    buildLocalReference: (json: any, baseUri: any, fragment: string, i: number) => string
```
Comment (darkgl0w) [inline index.d.ts:123]:
```suggestion
    /** Clone the input schema without changing it. Default to `false`. */
    clone?: boolean;
    buildLocalReference: (
      /** The `json` that is being resolved. */
      json: any,
      /** The `baseUri` object of the schema. */
      baseUri: {
        scheme?: string;
        userinfo?: string;
        host?: string;
        port?: number | string;
        path?: string;
        query?: string;
        fragment?: string;
        reference?: string;
        error?: string;
      },
      /** `fragment` is the `$ref` string when the `$ref` is a relative reference. */
      fragment: string,
      /** `i` is a local counter to generate a unique key. */
      i: number
    ) => string;
```

It was missing the clone option, and baseUri is the parse result from `uri-js`.

**Review by darkgl0w — COMMENTED** [review]
Nitpicks.

the `json` property can be typed as `Record<string, any>` I think. Can you explore this way ?

EDIT: Maybe we can type JSON Objects like this :

```ts
type JSONValue = boolean | Array<JSONValue> | number | null | string | Record<string, JSONValue>;

interface JSONObject extends Record<string, JSONValue> {}
```

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by samzilverberg** [comment]
@darkgl0w re using a `JSONObject` :

lib users that need to use `refResolver` would have to go the the documentation and figure out what `json` is and what fields it contains (it's a pretty flexible schema json AFAIK).
The compiler and the added `JSONObject` types are not going to be much help with that.
I think `any` is a good first step and that a more accurate typing can be added later from someone that has better knowledge of `refResolver` (or has more time to explore).

**Comment by darkgl0w** [comment]
@samzilverberg I gave you the typescript definition for `JSONObject`. You just need to add this :
```ts
export type JSONValueType =
  | boolean
  | JSONObject
  | JSONValueType[]
  | null
  | number
  | string;
export interface JSONObject {
  [key: string]: JSONValueType;
}
```

and then change the `json` type from `any` to `JSONObject` as it covers all possible values. :smiley: 
With this users can type their `json` objects by importing the `JSONObject` interface and if needed they can also import the `JSONValue` type. ^^

**Comment by samzilverberg** [comment]
i added a json type for the `json` field

File: index.d.ts [hunk 2]
```diff
@@ -118,6 +125,31 @@ export interface FastifyDynamicSwaggerOptions extends FastifySwaggerOptions {
    * Overwrite the route schema
    */
   transform?: Function;
+
+  refResolver?: {
+    /** Clone the input schema without changing it. Default to `false`. */
+    clone?: boolean;
+    buildLocalReference: (
+      /** The `json` that is being resolved. */
+      json: JSONValue,
```
Comment (darkgl0w) [inline index.d.ts:134]: `'hello'`, `1`, `true` ... etc are valid according to this type.

**Review by darkgl0w — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 2 above)
Comment (samzilverberg) [reply index.d.ts:134]: ahh i see. i'll add the JSONObject type as well :)

File: index.d.ts [hunk 3]
```diff
@@ -103,6 +103,17 @@ export type FastifySwaggerInitOAuthOptions = Partial<{
   usePkceWithAuthorizationCodeGrant: boolean
 }>
 
+type JSONValue =
+    | string
+    | number
```
Comment (darkgl0w) [inline index.d.ts:108]:
```suggestion
    | null
    | number
```
`null` is a valid type too for a json object ^^

**Review by darkgl0w — APPROVED** [review]
LGTM with the suggested change. ^^

File: index.d.ts (hunk 3 above)
Comment (samzilverberg) [reply index.d.ts:108]: ty for spotting that. the trouble with making changes late in the evening is you (me) miss the small details

**Comment by samzilverberg** [comment]
@darkgl0w  / @mcollina i'm not sure if I can merge or if you have to do it it.
other than that, github shows me a warning saying that a maintainer (one of you?) needs to approve a workflow.

**Review by climba03003 — APPROVED** [review]

**Pull request merged**